ARCs/arc-0032.md

```diff
@@ -0,0 +1,233 @@
+
+### Definitions
+
+- [Application Specification](#application-specification): The object containing the elements describing the Application.
+- [Source Specification](#source-specification): The object containing a description of the TEAL source programs that are evaluated when this Application is called.
+- [Schema Specification](#schema-specification): The object containing a description of the schema required by the Application.
+- [Type Specification](#type-specification): The object containing a map type name of name to the user defined data types
+- [Error Specification](#error-specification): The object containing a map of `pc` to Error message
+
+
+### Application Specification
+
+The Application Specification is composed of a number of elements that serve to fully describe the Application. 
+
+```ts
+type AppSpec = {
+  // embedded contract fields, see ARC-0004 for more
+  ABIContract;
+  // the original teal source, containing annotations, base64 encoded
+  source?: SourceSpec;
+  // the schema this application requires/provides
+  schema?: SchemaSpec;
+  // map of type name=>user defined type
+  types?: TypeSpec;
+  // map of pc=>error message
+  errors?: ErrorSpec;
+}
+```
+### Source Specification
+
+Contains the source TEAL files including comments and other annotations.
+
+```ts
+// Opject containing the original TEAL source files
+type SourceSpec = {
+  // b64 encoded approval program
+  approval: string;
+  // b64 encoded clear state program
+  clear: string;
+}
+```
+
+### Schema Specification
+
+TODO: describe diff between Declared/Reserved and motivation for providing both
+TODO: describe how to decode the type and the value given the decoded type
+
+The schema of an application is critical to know prior to creation since it is immutable after create. It also helps clients of the application understand the data that is available to be queried from off chain. Individual fields can be referenced from the [default arguments](#default-arguments) to provide input data to a given ABI method.
+
+While some fields are possible to know ahead of time, others may be keyed dynamically. In both cases the data type being stored MUST be known and declared ahead of time. 
+
+```ts
+// The complete schema for this application
+type SchemaSpec = {
+  local: Schema;
+  global: Schema;
+}
+
+// Schema fields may be declared explicitly or reserved
+type Schema = {
+  declared: Record<string, DeclaredSchemaValueSpec>;
+  reserved: Record<string, ReservedSchemaValueSpec>;
+}
+
+// Types supported for encoding/decoding
+enum AVMType { uint64, bytes }
+// string encoded datatype name defined in arc-4
+type ABIType = string;
+
+// Fields that have an explicit key
+type DeclaredSchemaValueSpec = {
+  type: AVMType | ABIType;
+  key: string;
+  desc: string;
+}
+
+// Fields that have an undetermined key
```

Comment:
What about the case of known keys, but they're autogenerated from some prefix. E.g. if you have `blob_0`, `blob_1`, and so on. These keys are determined already, but would be unwieldy to list individually?